### PR TITLE
client: fix FailOnNonTempDialError and add a test for it

### LINKED ISSUE #2266: FailOnNonTempDialError doesn't fail
### What version of gRPC are you using?

07ef407d991f1004e6c3367c8f452ed9a02f17ff

### What version of Go are you using (`go version`)?

`go version go1.10.3 darwin/amd64`

### What did you do?
If possible, provide a recipe for reproducing the error.

```
package main

import (
	"context"
	"fmt"
	"google.golang.org/grpc"
	"net"
	"time"
)

type fatalError struct{}

func (fatalError) Error() string   { return "fatal" }
func (fatalError) Temporary() bool { return false }

func main() {

	dialer := grpc.WithDialer(func(string, time.Duration) (net.Conn, error) {
		fmt.Println("Fail")
		return nil, fatalError{}
	})

	conn, err := grpc.DialContext(context.Background(), "nope.server:8080",
		grpc.WithInsecure(),
		grpc.WithBlock(),
		grpc.FailOnNonTempDialError(true),
		dialer,
	)

	fmt.Printf("got %v %v\n", conn, err)
}
```

### What did you expect to see?

The documentation for FailOnNonTempDialError says:

> FailOnNonTempDialError returns a DialOption that specifies if gRPC fails on non-temporary dial errors. If f is true, and dialer returns a non-temporary error, gRPC will fail the connection to the network address and won't try to reconnect. 

I would expect that if a non-temporary error was returned from the dialer, I would see the dial immediately return.

```
Fail
got <something>fatal
```

### What did you see instead?

Instead, it just prints `Fail` over and over again since it keeps trying to retry (the thing it's not supposed to do).

```
Fail
Fail
Fail
...
```

Perhaps I'm misunderstanding the documentation?

### PR DESCRIPTION
**Pull request by virtuald**
- Fixes #2266

### CODE REVIEW

**Comment by virtuald** [comment]
Added CLA.

File: clientconn_test.go [hunk 1]
```diff
@@ -443,6 +444,25 @@ func TestDialContextCancel(t *testing.T) {
...
 }
 
+type failFastError struct{}
+
+func (failFastError) Error() string   { return "failfast" }
+func (failFastError) Temporary() bool { return false }
+
+func TestDialContextFailFast(t *testing.T) {
+	defer leakcheck.Check(t)
+	ctx := context.Background()
```
Comment (dfawley) [inline clientconn_test.go]: Please make this timeout eventually so the test won't hang forever if it breaks.  E.g.

```go
ctx := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment (virtuald) [reply clientconn_test.go]: Why don't any of the other tests do that then?

**Comment by dfawley** [comment]
> Also, thinking about it, should a note be added to the documentation that FailOnNonTempDialError is only useful if WithBlock is also used? It's not immediately clear to me that it does anything if not using WithBlock...

That would be great, thanks.

The whole idea of a blocking `Dial` is a bit unusual for the gRPC model, where a `ClientConn` is actually a long-lived pool of connections that may be unavailable at times.  The initial connection attempt is only interesting in that if it fails, it's possible your configuration was incorrect, vs. transient network/unavailability problems. 
 Note that in all other languages, gRPC does not attempt to connect until the first RPC is dispatched, and the initial connection is not treated specially.

**Comment by virtuald** [comment]
Added documentation note also.

**Pull request merged**